packages/util/src/liquidityProxy.ts

```diff
@@ -0,0 +1,1207 @@
+      market: LiquiditySourceTypes.XSTPool,
+    };
+  } else {
+    return {
+      result: tbcQuote(inputAssetId, outputAssetId, amount, isDesiredInput, payload),
+      market: LiquiditySourceTypes.MulticollateralBondingCurvePool,
+    };
+  }
+};
+
+// xor is output
+const primaryMarketAmountBuyingXor = (
+  collateralAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  xorReserve: FPNumber,
+  otherReserve: FPNumber,
+  payload: QuotePayload
+): FPNumber => {
+  try {
+    const secondaryPrice = FPNumber.isGreaterThan(xorReserve, FPNumber.ZERO)
+      ? safeDivide(otherReserve, xorReserve)
+      : MAX;
+
+    const primaryBuyPrice =
+      collateralAssetId === XSTUSD
+        ? xstBuyPriceNoVolume(collateralAssetId, payload)
+        : tbcBuyPriceNoVolume(collateralAssetId, payload);
+
+    const k = xorReserve.mul(otherReserve);
+
+    if (isDesiredInput) {
+      if (FPNumber.isLessThan(secondaryPrice, primaryBuyPrice)) {
+        const amountSecondary = k.mul(primaryBuyPrice).sqrt().sub(otherReserve);
+
+        if (FPNumber.isGreaterThanOrEqualTo(amountSecondary, amount)) {
+          return FPNumber.ZERO;
+        } else if (FPNumber.isLessThanOrEqualTo(amountSecondary, FPNumber.ZERO)) {
+          return amount;
+        } else {
+          return amount.sub(amountSecondary);
+        }
+      } else {
+        return amount;
+      }
+    } else {
+      if (FPNumber.isLessThan(secondaryPrice, primaryBuyPrice)) {
+        const amountSecondary = xorReserve.sub(safeDivide(k, primaryBuyPrice).sqrt());
+
+        if (FPNumber.isGreaterThanOrEqualTo(amountSecondary, amount)) {
+          return FPNumber.ZERO;
+        } else if (FPNumber.isLessThanOrEqualTo(amountSecondary, FPNumber.ZERO)) {
+          return amount;
+        } else {
+          return amount.sub(amountSecondary);
+        }
+      } else {
+        return amount;
+      }
+    }
+  } catch (error) {
+    return FPNumber.ZERO;
+  }
+};
+
+// xor is input
+const primaryMarketAmountSellingXor = (
+  collateralAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  xorReserve: FPNumber,
+  otherReserve: FPNumber,
+  payload: QuotePayload
+): FPNumber => {
+  try {
+    const secondaryPrice = FPNumber.isGreaterThan(xorReserve, FPNumber.ZERO)
+      ? safeDivide(otherReserve, xorReserve)
+      : FPNumber.ZERO;
+
+    const primarySellPrice =
+      collateralAssetId === XSTUSD
+        ? xstSellPriceNoVolume(collateralAssetId, payload)
+        : tbcSellPriceNoVolume(collateralAssetId, payload);
+
+    const k = xorReserve.mul(otherReserve);
+
+    if (isDesiredInput) {
+      if (FPNumber.isGreaterThan(secondaryPrice, primarySellPrice)) {
+        const amountSecondary = safeDivide(k, primarySellPrice).sqrt().sub(xorReserve);
+
+        if (FPNumber.isGreaterThan(amountSecondary, amount)) {
+          return FPNumber.ZERO;
+        } else if (FPNumber.isLessThanOrEqualTo(amountSecondary, FPNumber.ZERO)) {
```

Comment:
please add a function for `FPNumber.isLessThanOrEqualTo(something, FPNumber.ZERO)` inside this file